### PR TITLE
allow users to disable full text index for specific pages #1818

### PR DESCRIPTION
**Pull request by 928PJY**
#1818

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/ExtractSearchIndex.cs [hunk 1]
```diff
@@ -101,6 +101,9 @@ internal SearchIndexItem ExtractItem(HtmlDocument html, string href)
         {
             var contentBuilder = new StringBuilder();
 
+            if (html.DocumentNode.SelectNodes("//meta[contains(@content,'noindex')]") != null)
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/ExtractSearchIndex.cs]: >meta [](start = 49, length = 4)

select meta in head?

File: src/docfx.website.themes/default/partials/head.tmpl.partial [hunk 2]
```diff
@@ -14,6 +14,7 @@
   <link rel="stylesheet" href="{{_rel}}styles/main.css">
   <meta property="docfx:navrel" content="{{_navRel}}">
   <meta property="docfx:tocrel" content="{{_tocRel}}">
+  {{#_noindex}}<meta name="searchdisable" content="noindex">{{/_noindex}}
```
Comment (hellosnow) [inline src/docfx.website.themes/default/partials/head.tmpl.partial]: >searchdisable [](start = 27, length = 13)

how about "searchOption" ? "noindex" means disable already?

File: src/docfx.website.themes/default/partials/_head.liquid [hunk 3]
```diff
@@ -26,6 +26,9 @@
   <link rel="stylesheet" href="{{_rel}}styles/main.css">
   <meta property="docfx:navrel" content="{{_navRel}}">
   <meta property="docfx:tocrel" content="{{_tocRel}}">
+  {%- if _noindex -%}
+  <meta name="searchdisable" content="noindex">
```
Comment (vicancy) [inline src/docfx.website.themes/default/partials/_head.liquid]: naming: *Search* is enough?

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/ExtractSearchIndex.cs (hunk 1 above)
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/ExtractSearchIndex.cs]: `/html/head/meta[@meta="search"]

Comment (928PJY) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/ExtractSearchIndex.cs]: /html/head/meta[@name="search"]?

---
In reply to: [134167442](https://github.com/dotnet/docfx/pull/1983#discussion_r134167442) [](ancestors = 134167442)

File: src/docfx.website.themes/default/partials/head.tmpl.partial (hunk 2 above)
Comment (superyyrrzz) [inline src/docfx.website.themes/default/partials/head.tmpl.partial]: >searchdisable [](start = 27, length = 13)

or robots? [http://www.robotstxt.org/meta.html](http://www.robotstxt.org/meta.html)

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/ExtractSearchIndex.cs [hunk 4]
```diff
@@ -101,6 +101,9 @@ internal SearchIndexItem ExtractItem(HtmlDocument html, string href)
         {
             var contentBuilder = new StringBuilder();
 
+            if (html.DocumentNode.SelectNodes("//html/head/meta[contains(@content,'noindex')]") != null)
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/ExtractSearchIndex.cs]: use `/html`, `//` means starting from any child node, `/` means starting from root node

Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/ExtractSearchIndex.cs]: so `content=ahhanoindex` matches?...

File: test/Microsoft.DocAsCode.Build.Engine.Tests/ExtractSearchIndexFromHtmlTest.cs [hunk 5]
```diff
@@ -62,6 +62,30 @@ public void TestSearchableClass()
         }
 
         [Fact]
+        public void TestSearchDisableClass()
+        {
+            var rawHtml = @"
+<head>
+    <title>This is title in head metadata</title>
+    <meta name='searchdisable' content='noindex'>
```
Comment (vicancy) [inline test/Microsoft.DocAsCode.Build.Engine.Tests/ExtractSearchIndexFromHtmlTest.cs]: update ut

File: src/docfx.website.themes/default/partials/head.tmpl.partial [hunk 6]
```diff
@@ -14,6 +14,7 @@
   <link rel="stylesheet" href="{{_rel}}styles/main.css">
   <meta property="docfx:navrel" content="{{_navRel}}">
   <meta property="docfx:tocrel" content="{{_tocRel}}">
+  {{#_noindex}}<meta name="searchOption" content="noindex">{{/_noindex}}
```
Comment (vicancy) [inline src/docfx.website.themes/default/partials/head.tmpl.partial]: naming: Options, as you implements, content can be `"noindex, nofollow"` which contains multiple options?

Comment (928PJY) [inline src/docfx.website.themes/default/partials/head.tmpl.partial]: we won't support nofolloe

---
In reply to: [134405842](https://github.com/dotnet/docfx/pull/1983#discussion_r134405842) [](ancestors = 134405842)

File: test/Microsoft.DocAsCode.Build.Engine.Tests/ExtractSearchIndexFromHtmlTest.cs [hunk 7]
```diff
@@ -62,17 +66,45 @@ public void TestSearchableClass()
...
+            docfx can do anything...
+        </article>
+    </body>
+</html>
+";
+            var html = new HtmlDocument();
+            html.LoadHtml(rawHtml);
+            var href = "http://dotnet.github.io/docfx";
+            var item = _extractor.ExtractItem(html, href);
+            Assert.True(item == null);
```
Comment (superyyrrzz) [inline test/Microsoft.DocAsCode.Build.Engine.Tests/ExtractSearchIndexFromHtmlTest.cs]: >Assert.True [](start = 12, length = 11)

can use Assert.Null

**Comment by superyyrrzz** [comment]
update document later

File: src/Microsoft.DocAsCode.Build.Engine/PostProcessors/ExtractSearchIndex.cs [hunk 8]
```diff
@@ -101,6 +101,9 @@ internal SearchIndexItem ExtractItem(HtmlDocument html, string href)
         {
             var contentBuilder = new StringBuilder();
 
+            if (html.DocumentNode.SelectNodes("/html/head/meta[@name='searchOption' and @content='noindex']") != null)
+                return null;
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/PostProcessors/ExtractSearchIndex.cs]: wrap with `{}`

**Pull request merged**